[python] Output model to a pandas DataFrame

pford221 opened the pull request:
Added `trees_to_dataframe` method to Booster class in python API. Based on issue [#2578](https://github.com/microsoft/LightGBM/issues/2578).

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/LightGBM?pullRequest=2592) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/LightGBM?pullRequest=2592) before we can accept your contribution.<br/><br/>:x: pford221 [sign now](https://cla.opensource.microsoft.com/microsoft/LightGBM?pullRequest=2592)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/LightGBM?pullRequest=2592) it.</sub>

pford221 commented:
This is embarrassing. I left a python debugger `set_trace()` call in the unit test. Would you like me to resubmit the PR with it removed?

StrikerRUS commented:
You can simply add new commits in the `df_trees` branch and they will appear here.
And it seems that your new unit test fails itself. Please fix it along with PEP8 errors (you can find them here https://travis-ci.org/microsoft/LightGBM/jobs/616989890#L502). Thanks!

pford221 commented:
Hi,

I have formatted according to pep8 standard and the unit test I added is working.

StrikerRUS commented:
Tests fail due to bad indents in the file. One more reminder: you are **NOT** restricted to 80 chars line length. I guess it was the reason you decided to change indents.
https://github.com/microsoft/LightGBM/tree/master/python-package#development-guide

pford221 commented:
Hi. I appreciate your patience! I'm not too familiar with PEP 8 so I relied on the PEP 8 auto formatting python library and maybe it doesn't exactly conform to LightGBM's standards or maybe something happened in the copy and paste from the PEP 8 formatted .py file to basic.py. 

I went through all the travis-ci logs and corrected the formatting issues that I found. I then committed and squashed all the commits into one for easier review.

Thanks!

StrikerRUS commented:
No problem! Thank you very much for this PR! I'll review it shortly.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+
+        Returns
+        -------
+        Dataframe or list
```
```suggestion
        result : DataFrame or list
```

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+        Returns
+        -------
+        Dataframe or list
+            Returns a pandas Dataframe if pandas is installed. Otherwise returns a list.
```
```suggestion
            Returns a pandas DataFrame if pandas is installed, and a list otherwise.
```

pford221 replied:
Updated.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+            Returns a pandas Dataframe if pandas is installed. Otherwise returns a list.
+        """
+        if not PANDAS_INSTALLED:
+            print('Pandas is not installed and so the result will be returned as a list')
```
```suggestion
            warnings.warn('Pandas is not installed and so the result will be returned as a list')
```

pford221 replied:
Update.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+            print('Pandas is not installed and so the result will be returned as a list')
+
+        if self.num_trees() == 0:
+            print('There are no trees in this Booster and thus nothing to return')
```
```suggestion
            raise LightGBMError('There are no trees in this Booster and thus nothing to parse')
```

pford221 replied:
Updated.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+            print('There are no trees in this Booster and thus nothing to return')
+            return
+
+        def _is_split_node(t):
```
Please do not use one-letter variable names (`t`).

pford221 replied:
Changed `t` to `tree`.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+
+            def _get_node_index(t, tree_index):
+                tree_num = str(tree_index) + '-' if tree_index is not None else ''
+                node_type = 'S' if _is_split_node(t) else 'L'
```
Cache result from `_is_split_node(t)` in a variable.

pford221 replied:
Done.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                return feature_name
+
+            # Create the node record, and populate universal data members
+            node = {'tree_index': tree_index,
```
I don't see `internal_value`, `internal_weight`, `leaf_value` and `leaf_weight` fields in this dict.

pford221 replied:
I've added all of these. Note `internal_weight` and `leaf_weight` both map to `weight` in the output which I assume is the sum of the hessians.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                node['missing_type'] = t['missing_type']
+                node['internal_count'] = t['internal_count']
+            else:
+                node['leaf_weight'] = t['leaf_value']
```
Mistake here?

pford221 replied:
Not a mistake just confusing. I'm used to calling the predictions for a particular leaf (i.e. terminal node) the leaf weight. I've changed this to be consistent with LightGBM terminology.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                node['internal_count'] = t['internal_count']
+            else:
+                node['leaf_weight'] = t['leaf_value']
+                node['internal_count'] = t['leaf_count']
```
If you use this key for both `internal_count` from non-terminal nodes and `leaf_count` from leaves, then it's better to name it simply `count`.

pford221 replied:
Yes, I've done this. As mentioned above, I've also mapped both `leaf_weight` and `internal_weight` to `weight`.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                tree_num = str(tree_index) + '-' if tree_index is not None else ''
+                node_type = 'S' if _is_split_node(t) else 'L'
+                node_num = str(
+                    t['split_index' if _is_split_node(t) else 'leaf_index'])
```
We do not have 80-chars line limit, so there is no need to split this quite short line.

pford221 replied:
Changed.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                        tree_index=tree_index,
+                        feature_names=feature_names,
+                        parent_node=node['node_index'])
+                    # In tree format, "subtree_list"is a list of node records (dicts), and we
```
```suggestion
                    # In tree format, "subtree_list" is a list of node records (dicts), and we
```

pford221 replied:
Thanks for the typo fix.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                    res.extend(subtree_list)
+            return res
+
+        m_dict = self.dump_model()
```
`model_dict` for better readability. Or what did you mean under the `m` letter?

pford221 replied:
Done.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,115 @@ def free_network(self):
+                                                 feature_names=feature_names))
+
+        if PANDAS_INSTALLED:
+            return DataFrame(m_list, columns=m_list[0].keys())
```
`m_list[0].keys()` is not expected to be sorted and it can cause problems in different Python versions/implementations:
> Changed in version 0.25.0: If data is a list of dicts, column order follows insertion-order for Python 3.6 and later.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html

I guess you can simply omit `columns` argument and pandas will name columns properly by default.

pford221 replied:
I don't understand this suggestion. Not passing in the columns from `m_list[0].keys()` will cause the columns to be sorted alphabetically in Pandas versions prior to 0.25.0. Therefore, to ensure consistency in all versions of Pandas, we explicitly pass the order of the columns.

StrikerRUS replied:
Ah, I see! But now there is no particular order at all, as `m_list[0].keys()` is a set-like collection. Maybe it should be `sorted(m_list[0].keys())` to ensure that columns are always sorted alphabetically for any Python implementation and pandas version?

pford221 replied:
My intention is not to have the columns be alphabetically sorted even thought that would be consistent across versions of python using `sorted()` as you suggest. I think the ordering of the columns is intuitive to how one might think of a tree's hierarchy: `tree_index`, `node_index`, ...,leaf `count`. 

A coupe options: 
1.) For users with a pre-3.6 version of Python, their column ordering will be non-deterministic. I'll defer to you as a library maintainer to judge whether that is too unappealing. 
2.) Instead of using a `dict` , we use a `collections.OrderedDict` to preserve the order of the key,value pairs. Not sure if this will have any performance or other downstream effects, but it should make sure the ordering is consistent. 
3.) I hard-code the order of the columns with a non-dynamic list. I don't love this option because if dictionary keys in the `model_list` ever change, then we'll have to remember to update this static list of column names/dictionary keys.

StrikerRUS replied:
Yes, I love your way of ordering too. But we should be as much consistent as it possible. So, I personally see only №2 as a good workaround. Can we try it? I don't think that there will be any significant overhead in comparison to simple dict. If pandas is compatible with OrderedDict, then everything should be OK, I hope.

pford221 replied:
OK. I went with option #2. I'm import `OrderedDict()` at the top of basic.py. I hope that's OK.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,47 @@ def check_asserts(data):
+        data = lgb.Dataset(X, label=y)
+        params = {
+            "objective": "binary",
+            "metric": "auc",
```
Leave default metric because it isn't something that we test here.

pford221 replied:
Removed.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,47 @@ def check_asserts(data):
+            "min_data": 10,
+            "num_leaves": 15,
+            "verbose": -1,
+            "num_threads": 1,
```
Why is it needed here?

pford221 replied:
It's not. I thought these hyperparameters were standard across unit tests.

StrikerRUS replied:
Sorry for the confusion! My comment was addressed to `"num_threads": 1`. Anyway, other params are not necessary as well, and everything seems good now.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,47 @@ def check_asserts(data):
+        mod_split = bst.feature_importance('split')
+        mod_gains = bst.feature_importance('gain')
+        np.testing.assert_equal(tree_split, mod_split)
+        np.testing.assert_almost_equal(tree_gains, mod_gains)
```
```suggestion
        np.testing.assert_allclose(tree_gains, mod_gains)
```

https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose

pford221 replied:
Updated.

pford221 commented:
Thank you for the thorough review, Striker. I will make these edits in the next day or two!

pford221 commented:
Hello. I can't debug why the Travis CI build failed. It appears to be related to Too Many Requests network error. Please let me know if it's something in this PR that's causing the issue.

Thanks!

StrikerRUS commented:
Thank you for your updates! I'll review them soon.

Speaking about Travis, I'm sorry for making you confused with failed test. That test checks all URLs in our docs and is quite unstable due to obvious reasons. I've simply re-run it and now everything seems to be OK.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,119 @@ def free_network(self):
+
+        Returns
+        -------
+        DataFrame or list
```
```suggestion
        result : DataFrame or list
```

pford221 replied:
I will update this once we agree to what other changes are needed. :)

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,119 @@ def free_network(self):
+        Returns
+        -------
+        DataFrame or list
+            Returns a pandas Dataframe if pandas is installed, and a list otherwise.
```
```suggestion
            Returns a pandas DataFrame if pandas is installed, and a list otherwise.
```

pford221 replied:
I will update this once we agree to what other changes are needed. :)

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,119 @@ def free_network(self):
+                node['weight'] = tree['internal_weight']
+                node['count'] = tree['internal_count']
+            else:
+                node['leaf_value'] = tree['leaf_value']
```
Maybe merge `leaf_value` and `internal_value` into simple `value` as it was done for `count` and `weight`?

pford221 replied:
Hi. I can do this, but I have a couple concerns. The first is that I don't know what `internal_value` represents. I originally excluded it from the output of this method because I didn't think/know if it would be useful. Second, users will intuitively be able to interpret `leaf_value` column as the terminal node predictions for each tree, so they might be confused to find a single `value` column for every node and have trouble interpreting its meaning.

StrikerRUS replied:
https://github.com/microsoft/LightGBM/blob/d7f8aa53b8be5212887e97e474c90bc5d8beac19/include/LightGBM/tree.h#L395-L396
https://github.com/microsoft/LightGBM/blob/d7f8aa53b8be5212887e97e474c90bc5d8beac19/include/LightGBM/tree.h#L432-L434
I suppose that it is `leaf_value` from previous boosting stages. So I guess it should be merged, because it represents the same thing and now confuses even more with significant number of NaN values in the corresponding columns. With `S`/`L` encoding in `node_index` field I think that users will easily find actual leaf output values.

pford221 replied:
I have combined `leaf_value` and `internal_value` to be `value`. However, it seems like `internal_value` is on a different scale than `leaf_value`. Here's a screenshot of the output from the breast cancer dataset for a single tree model with `max_depth = 2`.

![image](https://user-images.githubusercontent.com/7775219/70870808-f7eec600-1f65-11ea-86d7-7e4d7816dbee.png)

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1853,119 @@ def free_network(self):
+        if PANDAS_INSTALLED:
+            return DataFrame(model_list, columns=model_list[0].keys())
+        else:
+            return model_list
```
Just a question: what are the cases when this kind of output is useful? To be a source for any other dataframe-like object or what for? In other words, in what terms it's better than raw output from `dump_model()`? I'm asking because it contradicts with the method name a little bit.

pford221 replied:
This is very convenient way to look at things like maximum and minimum split gains which is useful for selecting hyperparamters or ranges of hyperparameters such as `min_gain_to_split`. You would have to write a lot of custom code to that from `dump_model()` output, but with pandas it's very easy. Also, it's a convenient way to see how much trees are being pruned, how imbalanced they are, etc. Finally, it's useful for figuring out which variables tend to have "interaction" effects which can tell you something about your data. All of those things are much more difficult when the data is in nested key:value structure like from `dump.model()`.

StrikerRUS replied:
Thank you very much for providing a rationale of this PR, but I was referring to a case when there is no pandas installed: `return model_list`. Shouldn't we simply raise an error when pandas was not detected?

> ... but with pandas it's very easy.

pford221 replied:
Ah, I see. I agree, I don't think a list format would be useful at all. In the latest commit I'm not raising a LightGBM exception if pandas is not installed.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1854,116 @@ def free_network(self):
+        Returns
+        -------
+        result: DataFrame
+            Returns a pandas DataFrame of the parsed model
```
```suggestion
            Returns a pandas DataFrame of the parsed model.
```
nit.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1854,116 @@ def free_network(self):
+            Returns a pandas DataFrame of the parsed model
+        """
+        if not PANDAS_INSTALLED:
+            raise LightGBMError('This method cannot be run without Pandas installed')
```
```suggestion
            raise LightGBMError('This method cannot be run without pandas installed')
```
nit.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,35 @@ def check_asserts(data):
+                    imptcs.append(impcts_dict[col])
+                except KeyError:
+                    imptcs.append(0.)
+            return np.array(imptcs)
```
nit.
Please add a blank line after `return`.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,35 @@ def check_asserts(data):
+        mod_split = bst.feature_importance('split')
+        mod_gains = bst.feature_importance('gain')
+        np.testing.assert_equal(tree_split, mod_split)
+        np.testing.assert_allclose(tree_gains, mod_gains)
```
Can we have some more tests? For instance, we can check that there are exactly 10 trees in the DataFrame.

pford221 replied:
I added two more tests. One to make sure the node count in the top-level node of each of the trees is the same length as the data and the other to ensure we have 10 trees (which makes me a bit worried as I've seen whole trees get pruned but with `min_gain_to_split` at 0 in this test, we should be fine).

StrikerRUS replied:
Very awesome! Many thanks!

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,42 @@ def check_asserts(data):
+        np.testing.assert_equal(tree_split, mod_split)
+        np.testing.assert_allclose(tree_gains, mod_gains)
+        np.testing.assert_equal(num_trees_from_df, num_trees)
+        np.testing.assert_equal(obs_counts_from_df, len(y))
```
nit.
It's better to use `self.assertEqual()` for less overhead.

StrikerRUS commented:
For auto-closing mechanism: fixed #2578. Also I think it can be treated as a fix for #2320, because with this PR it would be possible to obtain actual values as following: `actual_min_gain_to_split = model.trees_to_dataframe()['split_gain'].min()`

jameslamb commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,42 @@ def check_asserts(data):
+        tree_gains = _imptcs_to_numpy(X, gains_dict)
+        mod_split = bst.feature_importance('split')
+        mod_gains = bst.feature_importance('gain')
+        num_trees_from_df = tree_df.tree_index.nunique()
```
could you change this to `tree_df['tree_index']` and replace all other `df.{column_name}` type references in a similar way?

This is safer and more future-proof, since `pandas` will not prevent you from creating a column whose name conflicts with a public attribute of `pd.DataFrame` but will prefer the attribute to the column when there is a conflict.

example:

```python
import pandas as pd
df = pd.DataFrame({'index': [1,2,3]})
df.index

# >>> RangeIndex(start=0, stop=3, step=1)
```

pford221 replied:
Sure. I just made this change and pushed.

jameslamb replied:
Looks great, thank you! I'll resolve this thread

StrikerRUS replied:
@jameslamb Please update your review status as well. Your requested changes block merging:
> Merging is blocked
Merging can be performed automatically once the requested changes are addressed.

StrikerRUS commented:
@jameslamb Were your comments fully addressed?

jameslamb commented:
> @jameslamb Were your comments fully addressed?

ah, totally missed the followup comment and commit in my GitHub emails!  Yes they were, thank you for the fix @pford221

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1854,116 @@ def free_network(self):
+
+        Returns
+        -------
+        result: DataFrame
```
```suggestion
        result : pandas DataFrame
```

pford221 replied:
Done in most recent (6th) commit.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1854,116 @@ def free_network(self):
         self.network = False
         return self
 
+    def trees_to_dataframe(self):
```
@pford221 I've just found a bug, please fix the following edge case:
```
import numpy as np
import lightgbm as lgb

X = np.ones((10, 2))
y = np.random.rand(10)
lgb_data = lgb.Dataset(X, y)

bst = lgb.train({}, lgb_data, num_boost_round=3)
bst.trees_to_dataframe()
```

```
KeyError                                  Traceback (most recent call last)
<ipython-input-5-86ad7722a0da> in <module>
----> 1 bst.trees_to_dataframe()

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in trees_to_dataframe(self)
   1961             model_list.extend(tree_dict_to_node_list(tree['tree_structure'],
   1962                                                      tree_index=tree['tree_index'],
-> 1963                                                      feature_names=feature_names))
   1964 
   1965         return DataFrame(model_list, columns=model_list[0].keys())

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in tree_dict_to_node_list(tree, node_depth, tree_index, feature_names, parent_node)
   1936                                       tree_index=tree_index,
   1937                                       feature_names=feature_names,
-> 1938                                       parent_node=parent_node)
   1939 
   1940             res = [node]

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in create_node_record(tree, node_depth, tree_index, feature_names, parent_node)
   1896             node['tree_index'] = tree_index
   1897             node['node_depth'] = node_depth
-> 1898             node['node_index'] = _get_node_index(tree, tree_index)
   1899             node['left_child'] = None
   1900             node['right_child'] = None

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in _get_node_index(tree, tree_index)
   1879                 is_split = _is_split_node(tree)
   1880                 node_type = 'S' if is_split else 'L'
-> 1881                 node_num = str(tree['split_index' if is_split else 'leaf_index'])
   1882                 return tree_num + node_type + node_num
   1883 

KeyError: 'leaf_index'
```

StrikerRUS replied:
I'm sorry, this comment had to go to the previous my review but GitHub lagged and it hadn't been appended there.

pford221 replied:
OK. This has been handled through a `try...expect` clause. Also, since a single node tree doesn't have a `count` or `weight` keys, I have to deal with this as well. I wrote a robust function to check if each leaf node comes from a single node tree.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,42 @@ def check_asserts(data):
+                try:
+                    imptcs.append(impcts_dict[col])
+                except KeyError:
+                    imptcs.append(0.)
```
I think it can be replaced with one-line solution `imptcs.append(impcts_dict.get(col, 0.))`

pford221 replied:
Changed in the final (6th) commit.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1854,122 @@ def free_network(self):
+
+            def _is_single_node_tree(tree):
+                tree_keys = [str(key) for key in tree.keys()]
+                return tree_keys == ['leaf_value']
```
I guess comparing sets directly will be more efficient and requires less code:
```
tree.keys() == {'leaf_value'}
```
And `keys` are already strings, so there is no need to cast them.

pford221 replied:
Thanks for the suggestion. I didn't know that the keys of a dictionary as a collection was a set!

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -328,3 +328,39 @@ def check_asserts(data):
+        def _imptcs_to_numpy(X, impcts_dict):
+            cols = ['Column_' + str(i) for i in range(X.shape[1])]
+            imptcs = []
+            for col in cols:
```
Now after removing `try`/`except` logic, it can be rewritten with list comprehension for better efficiency and readability:
```
imptcs = [impcts_dict.get(col, 0.) for col in cols]
```

pford221 replied:
Got it.

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1853,6 +1854,122 @@ def free_network(self):
+
+        Returns
+        -------
+        result: pandas DataFrame
```
Please fix formatting issue [I mentioned in the previous review](https://github.com/microsoft/LightGBM/pull/2592#discussion_r364430743):
```suggestion
        result : pandas DataFrame
```
